File: src/main/java/fi/quanfoxes/parser/patterns/CastPattern.java
package fi.quanfoxes.parser.patterns;

import java.util.List;

import fi.quanfoxes.lexer.OperatorToken;
import fi.quanfoxes.lexer.Operators;
import fi.quanfoxes.lexer.Token;
import fi.quanfoxes.lexer.TokenType;
import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Pattern;
import fi.quanfoxes.parser.Singleton;
import fi.quanfoxes.parser.nodes.CastNode;

public class CastPattern extends Pattern {
    public static final int PRIORITY = 19;

    private static final int OBJECT = 0;
    private static final int CAST = 1;
    private static final int TYPE = 2;

    public CastPattern() {
        // Pattern:
        // ... -> Type / Type.Subtype
        super(TokenType.FUNCTION | TokenType.IDENTIFIER | TokenType.NUMBER | TokenType.DYNAMIC, /* ... */
              TokenType.OPERATOR, /* -> */
              TokenType.IDENTIFIER | TokenType.DYNAMIC); /* Type / Type.Subtype */
    }

    @Override
    public int priority(List<Token> tokens) {
        return PRIORITY;
    }

    @Override
    public boolean passes(List<Token> tokens) {
        OperatorToken cast = (OperatorToken)tokens.get(CAST);
        return cast.getOperator() == Operators.CAST;
    }

	@Override
	public Node build(Context context, List<Token> tokens) throws Exception {
        Node object = Singleton.parse(context, tokens.get(OBJECT));
        Node type = Singleton.parse(context, tokens.get(TYPE));

        return new CastNode(object, type);
	}
}

File: Zigzag/Parser/Patterns/CastPattern.cs
using System.Collections.Generic;

public class CastPattern : Pattern
{
	public const int PRIORITY = 19;

	private const int OBJECT = 0;
	private const int CAST = 1;
	private const int TYPE = 2;

	// ... as Type
	public CastPattern() : base
	(
		TokenType.FUNCTION | TokenType.IDENTIFIER | TokenType.NUMBER | TokenType.DYNAMIC,
		TokenType.KEYWORD,
		TokenType.IDENTIFIER | TokenType.DYNAMIC
	) { }

	public override int GetPriority(List<Token> tokens)
	{
		return PRIORITY;
	}

	public override bool Passes(Context context, List<Token> tokens)
	{
		var cast = (KeywordToken)tokens[CAST];
		return cast.Keyword == Keywords.AS;
	}

	public override Node Build(Context context, List<Token> tokens)
	{
		var source = Singleton.Parse(context, tokens[OBJECT]);
		var type = Singleton.Parse(context, tokens[TYPE]);

		return new CastNode(source, type);
	}
}